### ISSUE TITLE
ScrollNotifications START/UPDATE/END triggered on each frame for Trackpad scroll

### ISSUE THREAD

**Issue by ltOgt** [flutter/flutter#88037]
## Steps to Reproduce
Create a new project on the latest master with the following `main.dart`:

<details>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    MaterialApp(
      title: 'Trackpad Scroll Events on each Frame',
      theme: ThemeData.dark(),
      home: Scaffold(
        body: Row(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            DefaultScrollView(),
            SizedBox(
              width: 100,
            ),
            CustomScrollView(),
          ],
        ),
      ),
    ),
  );
}

class DefaultScrollView extends StatelessWidget {
  const DefaultScrollView({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return NotificationListener<ScrollNotification>(
      onNotification: (scrollNotification) {
        if (scrollNotification is ScrollStartNotification) {
          print("START");
        } else if (scrollNotification is ScrollEndNotification) {
          print("END");
        }
        return false;
      },
      child: SingleChildScrollView(
        child: Column(
          children: [
            for (int i = 0; i < 100; i++)
              Container(
                width: 100,
                height: 100,
                color: (i % 2 == 0) ? Colors.blue : Colors.black,
                child: Center(
                  child: Text(
                    "$i",
                  ),
                ),
              ),
          ],
        ),
      ),
    );
  }
}

class CustomScrollView extends StatefulWidget {
  const CustomScrollView({
    Key? key,
  }) : super(key: key);

  @override
  State<CustomScrollView> createState() => _CustomScrollViewState();
}

class _CustomScrollViewState extends State<CustomScrollView> {
  bool isMoving = false;
  DateTime lastMoveTime = DateTime.now();
  static const lastMoveTimeWait = Duration(milliseconds: 100);

  void setupAndCheckLastScrollEvent() {
    if (isMoving == false) {
      print("START");
    }

    setState(() {
      isMoving = true;
      lastMoveTime = DateTime.now();
    });

    Future.delayed(lastMoveTimeWait).then((_) {
      if (!isMoving) return;

      final timeIfWasLast = lastMoveTime.add(lastMoveTimeWait);
      if (false == timeIfWasLast.isAfter(DateTime.now())) {
        setState(() {
          isMoving = false;
          print("END");
        });
      }
    });
  }

  late ScrollController sc;
  @override
  void initState() {
    sc = ScrollController();
    sc.addListener(setupAndCheckLastScrollEvent);
    super.initState();
  }

  @override
  void dispose() {
    sc.removeListener(setupAndCheckLastScrollEvent);
    sc.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      controller: sc,
      child: Column(
        children: [
          for (int i = 0; i < 100; i++)
            Container(
              width: 100,
              height: 100,
              color: (i % 2 == 0) ? Colors.blue : Colors.black,
              child: Center(
                child: Text(
                  "$i",
                ),
              ),
            ),
        ],
      ),
    );
  }
}

```

</details>

`DefaultScrollView` is a normal `SingleChildScrollView` wrapped with `NotificationListener<ScrollNotification>`.
Start and End events will print "START" and "END".

`CustomScrollView` is a proposal on how to fix the issue, more on that later.

_____

**Expected results:** 
"START" is printed when starting to scroll.
"END" is printed when ending to scroll.

**Actual results:** 
"START" and "END" are both printed on every frame (during scroll).

This is probably because `PointerScrollEvent` has no Start/Update/End variants.

______
Proposal:

`CustomScrollView` contains a timed callback that checks whether the last scroll event is some time ago, i.e. whether it is the last scroll event.
The first event triggers "START" the last event triggers "END":
```dart
bool isMoving = false;
  DateTime lastMoveTime = DateTime.now();
  static const lastMoveTimeWait = Duration(milliseconds: 100);

  void setupAndCheckLastScrollEvent() {
    if (isMoving == false) {
      print("START");
    }

    setState(() {
      isMoving = true;
      lastMoveTime = DateTime.now();
    });

    Future.delayed(lastMoveTimeWait).then((_) {
      if (!isMoving) return;

      final timeIfWasLast = lastMoveTime.add(lastMoveTimeWait);
      if (false == timeIfWasLast.isAfter(DateTime.now())) {
        setState(() {
          isMoving = false;
          print("END");
        });
      }
    });
  }
```

_____
doctor

<details>
  <summary>doctor</summary>

```
[✓] Flutter (Channel master, 2.5.0-6.0.pre.33, on Ubuntu 18.04.5 LTS 5.4.0-66-generic, locale en_US.UTF-8)
    • Flutter version 2.5.0-6.0.pre.33 at /home/omni/repos/flutter_web/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 089323487f (2 days ago), 2021-08-09 13:25:04 -0400
    • Engine revision 4fef55db10
    • Dart version 2.14.0 (build 2.14.0-385.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.0-rc2)
    • Android SDK at /home/omni/Android/Sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/linux#android-setup for more details.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /snap/bin/chromium

[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2
    • pkg-config version 0.29.1

[✓] Android Studio (version 3.5)
    • Android Studio at /opt/AppDev/SDK/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] VS Code
    • VS Code at /snap/code/current
    • Flutter extension version 3.25.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 18.04.5 LTS 5.4.0-66-generic
    • Chrome (web)    • chrome • web-javascript • Chromium 92.0.4515.131 snap

! Doctor found issues in 1 category.

```

</details>

Tested only on `Linux`, so this might not be an issue on `Mac` and `Windows`

P.S.
`InteractiveViewer` has the same issue for its `onInteractionStart/Update/End` callbacks

**Comment by ltOgt** [comment]
the output of `flutter doctor -v` is under the collapsed `doctor` detail.
here it is again:
```
[✓] Flutter (Channel master, 2.5.0-6.0.pre.33, on Ubuntu 18.04.5 LTS 5.4.0-66-generic, locale en_US.UTF-8)
    • Flutter version 2.5.0-6.0.pre.33 at /home/omni/repos/flutter_web/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 089323487f (2 days ago), 2021-08-09 13:25:04 -0400
    • Engine revision 4fef55db10
    • Dart version 2.14.0 (build 2.14.0-385.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.0-rc2)
    • Android SDK at /home/omni/Android/Sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/linux#android-setup for more details.

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /snap/bin/chromium

[✓] Linux toolchain - develop for Linux desktop
    • clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
    • cmake version 3.10.2
    • ninja version 1.8.2
    • pkg-config version 0.29.1

[✓] Android Studio (version 3.5)
    • Android Studio at /opt/AppDev/SDK/android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_202-release-1483-b49-5587405)

[✓] VS Code
    • VS Code at /snap/code/current
    • Flutter extension version 3.25.0

[✓] Connected device (2 available)
    • Linux (desktop) • linux  • linux-x64      • Ubuntu 18.04.5 LTS 5.4.0-66-generic
    • Chrome (web)    • chrome • web-javascript • Chromium 92.0.4515.131 snap

! Doctor found issues in 1 category.
```

I am aware of NotificationListener, and as you can see in the example I use it in `DefaultScrollView`, only on my machine the desired behaviour does not happen.
I assume it works on your machine, hence the confusion.

See attached video for the issue:

https://user-images.githubusercontent.com/24209580/129253400-01bdb442-e857-43fa-acbb-2230494b8157.mp4



So to recap:
`NotificationListener` triggers both start and end on each frame on my machine.
(This is the approach used in `DefaultScrollView`.
`CustomScrollView` is just a workaround and I proposed it since I assumed the issue might happened for all trackpads.

**Comment by maheshmnj** [comment]
@ltOgt, Thanks for the clarification. I am able to reproduce the issue. Looks like this issue is for Desktop and the web it works fine on mobile platforms (Android,Ios) 

<details>
<summary>Minimal reproducible code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(
    MaterialApp(
      title: 'Trackpad Scroll Events on each Frame',
      theme: ThemeData.dark(),
      home: Scaffold(
        body: DefaultScrollView(),
      ),
    ),
  );
}

class DefaultScrollView extends StatelessWidget {
  const DefaultScrollView({
    Key? key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return NotificationListener<ScrollNotification>(
      onNotification: (scrollNotification) {
        if (scrollNotification is ScrollStartNotification) {
          print("START");
        } else if (scrollNotification is ScrollUpdateNotification) {
          print("Update");
        } else if (scrollNotification is ScrollEndNotification) {
          print("END");
        }
        return false;
      },
      child: SingleChildScrollView(
        child: Column(
          children: [
            for (int i = 0; i < 100; i++)
              Container(
                // width: 100,
                height: 100,
                color: (i % 2 == 0) ? Colors.blue : Colors.black,
                child: Center(
                  child: Text(
                    "$i",
                  ),
                ),
              ),
          ],
        ),
      ),
    );
  }
}
```
</details>

<details>
<summary>flutter doctor -v</summary>

```console
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.5.2 20G95 darwin-arm, locale en-GB)
    • Flutter version 2.2.3 at /Users/mahesh/Documents/flutter
    • Framework revision f4abaa0735 (6 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (4 available)
    • iPhone 12 Pro Max (mobile) • 5430F628-A903-402C-A635-D55D7E6EEFCA • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 11.5.2
      20G95 darwin-arm
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome
      92.0.4515.131

• No issues found!
```

```console
[✓] Flutter (Channel master, 2.5.0-7.0.pre.8, on macOS 11.5.2 20G95 darwin-arm, locale en-GB)
    • Flutter version 2.5.0-7.0.pre.8 at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 40e8620a27 (7 days ago), 2021-08-06 09:14:50 +0430
    • Engine revision 431ac604da
    • Dart version 2.14.0 (build 2.14.0-383.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (4 available)
    • iPhone 12 Pro Max (mobile) • 5430F628-A903-402C-A635-D55D7E6EEFCA • ios            •
      com.apple.CoreSimulator.SimRuntime.iOS-14-5 (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 11.5.2
      20G95 darwin-arm
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome
      92.0.4515.131

• No issues found!
```
</details>

Thanks

**Comment by maheshmnj** [comment]
@ltOgt  I see that there is a similar issue https://github.com/flutter/flutter/issues/44732 describing this case, You may follow up there for further updates. Closing this issue as a duplicate if you disagree feel free to comment and I will reopen it.

Thank you.
